### ISSUE TITLE
ReceiveAsync By SequenceNumber is not supported in .net core for Azure Service Bus Subscriptions

### ISSUE THREAD

**Issue by evhfla** [Azure/azure-sdk-for-net#15097]
**Library or service name.**
Microsoft.Azure.ServiceBus
 
**Is your feature request related to a problem? Please describe.**
Need the ability to receive a specific message for a subscriber of a Topic by sequencenumber.  It is supported in the full .net sdk, but not supported in the .net core sdk.  The .net core sdk only supports PeekBySequenceNumberAsync, which doesn't acquire a lock and therefore, you can't complete that message.

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by jsquire** [comment]
//cc: @JoshLove-msft, @MiYanni

**Comment by SeanFeldman** [comment]
Note that this method was also "lost" in track 1 SDK while available in track 0.

**Comment by SeanFeldman** [comment]
Track 0 overload that takes `sequenceNumber` is [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.messagereceiver.receiveasync?view=azure-dotnet#Microsoft_ServiceBus_Messaging_MessageReceiver_ReceiveAsync_System_Int64_).

**Comment by axisc** [comment]
Discussed this with @SeanFeldman offline. While the functionality to receive with a sequence number isn't intuitive for a Queue or Subscription, it's adds value when peeking/receiving from a deadletter entity.

@evhfla did you have other scenarios in mind? Receiving from a standard queue/subscription with a sequence number is like querying a database so it's a less preferred pattern with queueing where the default should be FIFO.

**Comment by evhfla** [comment]
No, I only need it for the deadletter queue, as I want to give users the ability to choose which messages in the deadletter they want to retry, edit and retry, or just delete without retrying.

**Comment by axisc** [comment]
@evhfla, @SeanFeldman 
this feature was removed due to lack of support from the AMQP protocol (in track 0 it was implemented using SBMP). 

Since the use-case is for DLQs only, we will investigate internally and identify a possible solution or workaround.

**Comment by SeanFeldman** [comment]
> Since the use-case is for DLQs only, we will investigate internally and identify a possible solution or workaround.

Coincidentally, I have an example where this assumption does not hold true. A customer had a logical error and has sent messages with future enqueue date/time. Those messages need to be "patched". The process would include "receive by sequence number and replace with a correct message" for all messages that are impacted and scheduled. But at this point, it's not possible other than going with the old track 0 SDK. And this is not a DLQ, so the assumption is inaccurate, @axisc.

**Comment by ErikMogensen** [comment]
I am pretty sure [ReceiveAsync(Int64) in track 0](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.messagereceiver.receiveasync?view=azure-dotnet#Microsoft_ServiceBus_Messaging_MessageReceiver_ReceiveAsync_System_Int64_) only is able to receive deferred messages and the documentation should be updated. The ability to receive normal messages was something that I and others asked for a lot some years ago. Clemens was involved in at least one of the discussions.

Since it was important to be able to remove deadletter messages without having to start with the top one I implemented a nasty workaround for Service Bus Explorer where it peek-locks and abandons every message until it gets to the message which is to be removed. Since this increments the Delivery count for the messages that are older than the intended one, I only implemented this for the deadletter subqueue. 

So it would be great if this was implemented.

**Comment by paolosalvatori** [comment]
In the past, I submitted internal feedback several times and asked for the possibility to extend the possibility to receive messages from a queue or subscription by sequence number beyond deferred messages. In fact, this functionality is required to properly implement Message Repair and Resubmit pattern in the [Service Bus Explorer tool](https://github.com/paolosalvatori/ServiceBusExplorer) (Im' the owner of the tool, @SeanFeldman and @ErikMogensen are co-owners).

**Comment by aguetat** [comment]
@ErikMogensen could you share your solution plz (even if it's a bad workaround)

**Comment by ErikMogensen** [comment]
Sure @aguetat. It is in https://github.com/paolosalvatori/ServiceBusExplorer/blob/develop/src/Common/Helpers/DeadLetterMessageHandler.cs.

**Comment by awseeley** [comment]
I would also like the ability to be able to query the queue SequenceNumber. Our main scenario in particular is to be able to "fix" a particular message (E.g. if the body is incorrect for a downstream application).

The only workaround for this is to receive all messages until we get a match on the sequence number/message-id then take action to send the message back to the queue for processing. (As per @ErikMogensen 's workaround)

( I am using the azure-sdk-for-js which also has this limitation)

**Comment by ramya-rao-a** [comment]
@shankarsama, This feature would first need to be implemented in the service side. Can you please transfer/move this issue to the https://github.com/Azure/azure-service-bus repo where you collect feature requests for the service?

**Comment by SeanFeldman** [comment]
@kevinlam1 would it be possible to nudge internal folks? Got to admit that it's quite frustrating to see issues like this where broken internal communication is exposed.

**Comment by SeanFeldman** [comment]
@jsquire, do you not have the right to transfer the issue? And if you don't, could you raise a new one (I've seen folks doing that on behalf of the users raising issues in the .NET SDKs repo) rather than just backlogging?
I understand that it's not your job to chaise the messaging team and ensure they raise the issues. However, it would be helpful if issues don't just get buried in the backlog here but surface up. I appreciate your help.

**Comment by jsquire** [comment]
Hi @SeanFeldman.   Thanks for the ping;  the backlog assignment was part of a triage script that identified this issue because tagging was changed to a `feature-request` which requires a milestone.   This is the first time that I've looked in on this since the initial triage.

I just caught up on the thread;  I'm not sure why the transfer wasn't performed when it was identified as needed, but I'll take care of it now.
